tests/std/tests/P0896R4_ranges_alg_copy/test.cpp

```diff
@@ -9,51 +9,47 @@
-        auto result  = copy(basic_borrowed_range{input}, basic_borrowed_range{output}.begin());
-        STATIC_ASSERT(same_as<decltype(result),
-            copy_result<iterator_t<basic_borrowed_range<int const>>, iterator_t<basic_borrowed_range<int>>>>);
-        assert(result.in == basic_borrowed_range{input}.end());
-        assert(result.out == basic_borrowed_range{output}.end());
-        assert(ranges::equal(output, input));
-    }
-    { // Validate iterator + sentinel overload
-        int output[] = {-1, -1, -1};
-        basic_borrowed_range wrapped_input{input};
-        auto result = copy(wrapped_input.begin(), wrapped_input.end(), basic_borrowed_range{output}.begin());
-        STATIC_ASSERT(same_as<decltype(result),
-            copy_result<iterator_t<basic_borrowed_range<int const>>, iterator_t<basic_borrowed_range<int>>>>);
-        assert(result.in == wrapped_input.end());
-        assert(result.out == basic_borrowed_range{output}.end());
-        assert(ranges::equal(output, input));
-    }
-}
-
-int main() {
-    STATIC_ASSERT((smoke_test(), true));
-    smoke_test();
-}
+// Validate dangling story
+STATIC_ASSERT(
+    same_as<decltype(copy(borrowed<false>{}, static_cast<int*>(nullptr))), copy_result<ranges::dangling, int*>>);
+STATIC_ASSERT(same_as<decltype(copy(borrowed<true>{}, static_cast<int*>(nullptr))), copy_result<int*, int*>>);
 
 struct instantiator {
-    template <class In, class Out>
-    static void call(In&& in = {}, Out out = {}) {
-        (void) ranges::copy(in, std::move(out));
-        (void) ranges::copy(ranges::begin(in), ranges::end(in), std::move(out));
+    static constexpr int input[3] = {13, 42, 1729};
+    template <class In, class Write>
+    static constexpr void call() {
+        { // Validate iterator + sentinel overload
+            int output[3] = {-1, -1, -1};
+            In wrapped_input{input};
+            auto result = copy(wrapped_input.begin(), wrapped_input.end(), Write{output});
+            STATIC_ASSERT(same_as<decltype(result), copy_result<iterator_t<In>, Write>>);
+            assert(result.in == wrapped_input.end());
+            if constexpr (std::equality_comparable<Write>) {
+                assert(result.out == Write{output + 3});
+            }
+            assert(ranges::equal(output, input));
+        }
+        { // Validate range overload
+            int output[3] = {-1, -1, -1};
+            In wrapped_input{input};
+            auto result = copy(wrapped_input, Write{output});
+            STATIC_ASSERT(same_as<decltype(result), copy_result<iterator_t<In>, Write>>);
+            assert(result.in == wrapped_input.end());
+            if constexpr (std::equality_comparable<Write>) {
+                assert(result.out == Write{output + 3});
+            }
+            assert(ranges::equal(output, input));
+        }
     }
 };
 
-template void test_in_write<instantiator>();
+int main() {
+    STATIC_ASSERT((test_in_write<instantiator, const int, int>(), true));

```

Comment:
Would it make sense to default `test_in_write` to `const int` for the input range so we do not accidentally mutate it.

Reply:
I commented in the `copy_n` review that we should probably change all defaults for parameters used as element types of input iterators/ranges to `const int`. (Or maybe we should get rid of the defaults?)

Reply:
Nothing to do here I guess